state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1217,31 +1196,95 @@ pub mod pallet {
-			let (non_violating, violating): (Vec<_>, Vec<_>) =
-				swaps.into_iter().partition(|swap| {
-					let final_output = swap.final_output.unwrap();
-					swap.refund_params()
-						.as_ref()
-						.map_or(true, |params| final_output >= params.min_output)
-				});
+			return Ok(swaps)
+		}
 
-			if violating.is_empty() {
-				Ok(non_violating)
-			} else {
-				Err(BatchExecutionError::PriceLimitHit {
-					successful_swaps: non_violating.into_iter().map(|ctx| ctx.swap).collect(),
-					failed_swaps: violating.into_iter().map(|ctx| ctx.swap).collect(),
-				})
+		/// Attempts to find (and execute) a batch of swaps that wouldn't result in hitting the
+		/// price impact limit, starting with the given batch, and taking swaps out of the batch if
+		/// needed.
+		#[transactional]
+		fn execute_batch(
+			mut swaps: Vec<Swap<T>>,
+		) -> Result<BatchExecutionOutcomes<T>, DispatchError> {
+			let mut removed_swaps = vec![];
+
+			loop {
+				match Self::try_execute_within_price_impact_limit(swaps.clone()) {
+					Ok(executed_swaps) => {
+						// No liquidity failures, but we still need to check
+						// for FoK violations:
+
+						let (non_violating, violating): (Vec<_>, Vec<_>) =
+							executed_swaps.into_iter().partition(|swap| {
+								let final_output = swap.final_output.unwrap();
+								swap.refund_params()
+									.as_ref()
+									.map_or(true, |params| final_output >= params.min_output)
+							});
+
+						if violating.is_empty() {
+							break Ok(BatchExecutionOutcomes {
+								successful_swaps: non_violating,
+								failed_swaps: removed_swaps,
+							});
+						} else {
+							swaps = non_violating.into_iter().map(|ctx| ctx.swap).collect();
+							removed_swaps.extend(violating.into_iter().map(|ctx| ctx.swap));
+						}
+					},
+					Err(error) => {
+						let (asset, direction, amount, ref swap_states) = match error {
+							BatchExecutionError::SwapLegFailed {
+								asset,
+								direction,
+								amount,
+								swaps,
+							} => (asset, direction, amount, swaps),
+							BatchExecutionError::DispatchError { error } => return Err(error),
+						};
+
+						Self::deposit_event(Event::<T>::BatchSwapFailed {
+							asset,
+							direction,
+							amount,
+						});
+
+						// Find the largest swap from the failing pool/direction:
+						let (swap_to_remove, remaining_swaps) =
+							utilities::split_off_highest_impact_swap(swaps, swap_states, direction);
+
+						// If we couldn't find a swap to remove, or the remaining swaps are
+						// empty, abort; otherwise, remove the swap and try again with the
+						// remaining swaps:
+						match swap_to_remove {
+							Some(swap_to_remove) if !remaining_swaps.is_empty() => {
+								removed_swaps.push(swap_to_remove);
+								swaps = remaining_swaps;
+							},
+							_ => {
+								if let Some(swap_to_remove) = swap_to_remove {
+									removed_swaps.push(swap_to_remove);
```

Comment:
I moved `if remaining_swaps.is_empty()` right after `loop {` so that it is always checked (and prevents us from calling functions unnecessarily when we have 0 swaps to begin with). Looks much cleaner now.